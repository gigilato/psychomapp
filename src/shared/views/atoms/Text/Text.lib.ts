import { upperFirst, capitalize } from 'lodash'
import { ReactNode } from 'react'

import { TextTransform, TextVariant } from './Text.props'

export const textVariants: Record<TextVariant, string> = {
  // Title
  title: 'font-[SpaceGrotesk-Bold] text-[34px] leading-[42px]',
  subtitle: 'font-[SpaceGrotesk-Bold] text-[20px] leading-[28px]',
  // Body
  body: 'font-[SpaceGrotesk-Regular] text-[14px] leading-[18px]',
  boldBody: 'font-[SpaceGrotesk-Medium] text-[14px] leading-[18px]',
}

export const getContent = (children: ReactNode, textTransform?: TextTransform) => {
  if (typeof children !== 'string' || !textTransform) return children
  const text = children ?? ''
  switch (textTransform) {
    case 'lowercase':
      return text.toLowerCase()
    case 'uppercase':
      return text.toUpperCase()
    case 'capitalize':
      return capitalize(text)
    default:
      return upperFirst(text)
  }
}
