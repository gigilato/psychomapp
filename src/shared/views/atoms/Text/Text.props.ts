import { TextProps } from 'react-native'

export type TextTransform = 'uppercase' | 'lowercase' | 'upper-first' | 'capitalize'
export type TextVariant =
  | 'title'
  | 'subtitle'
  // Body
  | 'body'
  | 'boldBody'
  | 'italicBody'

export interface ITextProps extends Omit<TextProps, 'textTransform'> {
  textTransform?: TextTransform
  variant?: TextVariant
  reversed?: boolean
}
