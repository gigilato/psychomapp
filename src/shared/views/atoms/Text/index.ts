export * from './Text'
export * from './Text.props'
export { textVariants } from './Text.lib'
