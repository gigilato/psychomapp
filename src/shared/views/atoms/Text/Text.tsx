/* eslint-disable @typescript-eslint/strict-boolean-expressions */
import clsx from 'clsx'
import { styled } from 'nativewind'
import React, { Component } from 'react'
import { Text as TextDefault } from 'react-native'
import Animated from 'react-native-reanimated'

import { getContent, textVariants } from './Text.lib'
import { ITextProps } from './Text.props'

const StyledText = styled(TextDefault)

export const Text = ({
  variant = 'body',
  className,
  textTransform,
  children,
  reversed,
  ...props
}: ITextProps) => {
  return (
    <StyledText
      className={clsx(textVariants[variant], reversed && 'text-white-light', className)}
      {...props}
    >
      {getContent(children, textTransform)}
    </StyledText>
  )
}

class TextComponent extends Component<ITextProps> {
  render() {
    return <Text {...this.props} />
  }
}
export const AnimatedText = Animated.createAnimatedComponent<ITextProps>(TextComponent)
