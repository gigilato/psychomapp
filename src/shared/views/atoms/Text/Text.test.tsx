import { render } from '@testing-library/react-native'

import { Text } from './Text'

describe('Text', () => {
  it('renders correctly', async () => {
    const text = '1st test'
    const component = render(<Text>{text}</Text>)
    await component.findByText(text)
    expect(component).toMatchSnapshot()
  })

  it('renders with variant', async () => {
    const text = '1nd test'
    const component = render(<Text variant="title">{text}</Text>)
    await component.findByText(text)
    expect(component).toMatchSnapshot()
  })

  it('renders with overrided styles', async () => {
    const text = '3rd test'
    const className = 'text-primary-classic text-[20px]'
    const component = render(
      <Text variant="title" className={className}>
        {text}
      </Text>
    )
    await component.findByText(text)
    expect(component).toMatchSnapshot()
  })
})
