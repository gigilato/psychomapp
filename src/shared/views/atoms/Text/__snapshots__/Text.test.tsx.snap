// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text renders correctly 1`] = `
<Text
  style={
    [
      {
        "fontFamily": "SpaceGrotesk-Regular",
      },
      {
        "fontSize": 14,
      },
      {
        "lineHeight": 18,
      },
    ]
  }
>
  1st test
</Text>
`;

exports[`Text renders with overrided styles 1`] = `
<Text
  style={
    [
      {
        "fontFamily": "SpaceGrotesk-Bold",
      },
      {
        "fontSize": 34,
      },
      {
        "lineHeight": 42,
      },
      {
        "color": "#6C464F",
      },
      {
        "fontSize": 20,
      },
    ]
  }
>
  3rd test
</Text>
`;

exports[`Text renders with variant 1`] = `
<Text
  style={
    [
      {
        "fontFamily": "SpaceGrotesk-Bold",
      },
      {
        "fontSize": 34,
      },
      {
        "lineHeight": 42,
      },
    ]
  }
>
  1nd test
</Text>
`;
